### PR TITLE
Feat: Add Dictionaries

### PR DESCRIPTION
**Pull request by rodrigo-pino**
Solves #138 

Add all dictionary related hints. I had to implement them all together to have a good grasp of all the context scopes. I didn't add tests on purpose to avoid re-writing any test in the case I'd misunderstood something down the line.

This means that dictionaries are implemented but tests are lacking, they'll come as follow up, together with benchmark tests.

This includes the following hints for dictionaries:

* AllocFelt252Dict
* Felt5252DictEntryInit
* Felt252DictEntryUpdate
* GetSegmentArenaIndex

And the following hints for squashed dictionaries:

* InitSquashData
* GetCurrentAccessIndex
* ShouldSkipSquashLoop
* GetCurrentAccessDelta
* ShouldContinueSquashLoop
* GetNextDictKey

Both dictionaries and squashed dictionaries come with their own set of execution contexts:

- Dictionaries use the context to keep track of what dictionary is in what segment and the values it holds given a key
- Squashed Dictionaries keeps track of the keys that have been already been checked out

### CODE REVIEW

File: pkg/hintrunner/hint_test.go [hunk 1]
```diff
@@ -184,8 +185,14 @@ func TestWideMul128(t *testing.T) {
 	var dstLow ApCellRef = 1
 	var dstHigh ApCellRef = 2
 
-	lhs := Immediate(*big.NewInt(1).Lsh(big.NewInt(1), 127))
-	rhs := Immediate(*big.NewInt(1<<8 + 1))
+	lhsBytes := uint256.NewInt(1).Lsh(uint256.NewInt(1), 127).Bytes32()
```
Comment (jmjac) [inline pkg/hintrunner/hint_test.go:188]: I think it's clearer as `new(uint256.Int).Lsh(uint256.NewInt(1), 127).Bytes32()` given that `newInt(1)` is instantly overwritten

File: pkg/hintrunner/hint_test.go [hunk 2]
```diff
@@ -220,8 +227,12 @@ func TestWideMul128IncorrectRange(t *testing.T) {
 	var dstLow ApCellRef = 1
 	var dstHigh ApCellRef = 2
 
-	lhs := Immediate(*big.NewInt(1).Lsh(big.NewInt(1), 128))
-	rhs := Immediate(*big.NewInt(1))
+	lhsBytes := uint256.NewInt(1).Lsh(uint256.NewInt(1), 128).Bytes32()
```
Comment (jmjac) [inline pkg/hintrunner/hint_test.go:230]: `new(uint256.Int).Lsh(uint256.NewInt(1), 128).Bytes32()`

File: pkg/hintrunner/hint.go [hunk 3]
```diff
@@ -287,13 +289,362 @@ func (hint SquareRoot) Execute(vm *VM.VirtualMachine) error {
...
+	return "GetCurrentAccessDelta"
+}
+
+func (hint *GetCurrentAccessDelta) Execute(vm *VM.VirtualMachine, ctx *HintRunnerContext) error {
+	indexDeltaPtr, err := hint.IndexDeltaMinusOne.Get(vm)
+	if err != nil {
+		return fmt.Errorf("get index delta address: %w", err)
+	}
+
+	previousKeyIndex := ctx.SquashedDictionaryManager.PopIndex()
```
Comment (jmjac) [inline pkg/hintrunner/hint.go:599]: This will panic if the keys list is empty. Is that situation possible?

File: pkg/hintrunner/hintrunner.go [hunk 4]
```diff
@@ -4,25 +4,215 @@ import (
...
+		idx:  uint64(len(dm.dictionaries)),
+	}
+	return newDictAddr
+}
+
+// Given a memory address, it looks for the right dictionary using the segment index. If no
+// segment is associated with the given segment index, it errors
+func (dm *DictionaryManager) GetDictionary(dictAddr *mem.MemoryAddress) (Dictionary, error) {
+	dict, ok := dm.dictionaries[dictAddr.SegmentIndex]
+	if !ok {
```
Comment (jmjac) [inline pkg/hintrunner/hintrunner.go:65]: nit: All other functions in this file follow pattern 
```
if dict, ok := ...; ok{
     return value;
}
return error;
```
this one is the only that does the inverse

File: pkg/hintrunner/hint.go [hunk 5]
```diff
@@ -371,13 +373,383 @@ func (hint SquareRoot) Execute(vm *VM.VirtualMachine) error {
...
+
+	value, err := hint.Value.Resolve(vm)
+	if err != nil {
+		return fmt.Errorf("resolve value: %w", err)
+	}
+
+	return ctx.DictionaryManager.Set(&dictPtr, key, &value)
+}
+
+type GetSegmentArenaIndex struct {
```
Comment (jmjac) [inline pkg/hintrunner/hint.go:513]: Why is this struct the only one that's `String()` value is different than struct name?

**Review by jmjac — APPROVED** [review]

File: pkg/hintrunner/hint.go (hunk 5 above)
Comment (rodrigo-pino) [reply pkg/hintrunner/hint.go:513]: oops!

File: pkg/hintrunner/constants.go
Comment (jkktom) [inline pkg/hintrunner/constants.go:1]: @rodrigo-pino Can you please let me know why was this file deleted?

Comment (rodrigo-pino) [reply pkg/hintrunner/constants.go:1]: It was moved

Comment (rodrigo-pino) [reply pkg/hintrunner/constants.go:1]: It's new location makes all constant accessible to any project pkg. We had some repetition around
